src/dune_rules/env_node.ml

```diff
@@ -15,90 +15,97 @@ end
+let menhir_flags t = Memo.Lazy.Async.force t.menhir_flags
 
-let format_config t = Memo.Lazy.force t.format_config
+let format_config t = Memo.Lazy.Async.force t.format_config
 
 let set_format_config t format_config =
-  { t with format_config = Memo.Lazy.of_val format_config }
+  { t with format_config = Memo.Lazy.Async.of_val format_config }
 
-let odoc t = Memo.Lazy.force t.odoc
+let odoc t = Memo.Lazy.Async.force t.odoc
 
-let coq t = Memo.Lazy.force t.coq
+let coq t = Memo.Lazy.Async.force t.coq
 
 let make ~dir ~inherit_from ~scope ~config_stanza ~profile ~expander
     ~expander_for_artifacts ~default_context_flags ~default_env
     ~default_bin_artifacts =
+  let open Memo.Build.O in
   let config = Dune_env.Stanza.find config_stanza ~profile in
   let inherited ~field ~root extend =
-    Memo.lazy_ (fun () ->
-        extend
-          (match inherit_from with
-          | None -> root
-          | Some t -> field (Memo.Lazy.force t)))
+    Memo.lazy_async (fun () ->
+        (match inherit_from with
+        | None -> Memo.Build.return root
+        | Some t -> Memo.Lazy.Async.force t >>= field)
+        >>= extend)
   in
   let inherited_if_absent ~field ~root f_absent =
-    Memo.lazy_ (fun () ->
+    Memo.lazy_async (fun () ->
         match root with
-        | None ->
-          f_absent
-            (match inherit_from with
-            | None -> None
-            | Some t -> Some (field (Memo.Lazy.force t)))
-        | Some x -> x)
+        | None -> (
+          match inherit_from with
+          | None -> f_absent None
+          | Some t ->
+            let* t = Memo.Lazy.Async.force t in
+            let* field = field t in
+            f_absent (Some field))
+        | Some x -> Memo.Build.return x)
   in
   let local_binaries =
     inherited ~field:local_binaries ~root:[] (fun binaries ->
-        binaries
-        @ List.map config.binaries
+        let+ expanded =
+          Memo.Build.sequential_map config.binaries
             ~f:
-              (File_binding.Unexpanded.expand ~dir ~f:(fun template ->
-                   Expander.Static.expand_str
-                     (Memo.Lazy.force expander_for_artifacts)
-                     template)))
+              (File_binding.Unexpanded.expand_build ~dir ~f:(fun template ->
+                   let+ expander_for_artifacts =
+                     Memo.Lazy.Async.force expander_for_artifacts
+                   in
+                   Expander.Static.expand_str expander_for_artifacts template))
+        in
+        binaries @ expanded)
   in
   let external_env =
     inherited ~field:external_env ~root:default_env (fun env ->
         let env, have_binaries =
           (Env.extend_env env config.env_vars, List.is_non_empty config.binaries)
         in
-        if have_binaries then
-          let dir = Utils.local_bin dir |> Path.build in
-          Env.cons_path env ~dir
-        else
-          env)
+        Memo.Build.return
```

Comment:
Here the difference is small, but you could push down this call as well.

Reply:
Done.